### PR TITLE
Chore/migrate pages

### PR DESCRIPTION
**Pull request by thatbudakguy**
this PR serves as a record of work done towards the 3.0 milestone on two issues:
- #188 
- #180 

the migration utilities provided by #188 are a prerequisite to complete all other work in this milestone. view the checklist on #188 for a record of work that has been done up to this point implementing and testing these utilities.

the current focus is a fully-functioning `add_revision` function that can call `html_to_streamfield` to create a page revision with arbitrary html content stored as a `StreamField`. this will allow migrating mezzanine pages to the new content structure, which is the focus of #180.

once this is done, migrations can be written to automatically find and migrate the home page, landing pages, and all content pages (in that order). see [0002_homepage.py](https://github.com/Princeton-CDH/cdh-web/blob/chore/migrate-pages/cdhweb/pages/migrations/0002_homepage.py) for an incomplete example of a migration for the homepage.

in addition to the incomplete functions & tests, one known bug is visible when testing `add_child()`: for some reason the landing page in the fixture is treated as (`instanceof`) `HomePage` in addition to `LandingPage`. it's not obvious why this is the case since the content types in the fixture seem correct.

### CODE REVIEW

File: cdhweb/pages/migration_utils.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+from django.core.serializers import serialize
+from django.db import connection
+from django.db.migrations.executor import MigrationExecutor
+from django.test import TestCase
+
+logger = logging.getLogger('wagtail.core')
+
+# see for an explanation of django-treebeard & the `path` attribute:
+# http://www.agilosoftware.com/blog/django-treebard-and-wagtail-page-creation/
+def get_parent(apps, page):
```
Comment (rlskoeser) [inline cdhweb/pages/migration_utils.py:14]: Why is this necessary? Don't pages have a `get_parent` method?

File: cdhweb/pages/migration_utils.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    # otherwise remove the final 4 digits to get the parent's id (pk)
+    Page = apps.get_model('wagtailcore', 'Page')
+    return Page.objects.get(path=page.path[:-4])
+
+# add_child utility method adapted from:
+# http://www.agilosoftware.com/blog/django-treebard-and-wagtail-page-creation/
+# see also:
+# https://github.com/wagtail/wagtail/blob/stable/2.3.x/wagtail/core/models.py#L442
+def add_child(apps, parent_page, klass, **kwargs):
```
Comment (rlskoeser) [inline cdhweb/pages/migration_utils.py:31]: Same question about this one as `get_parent`. The docs mention a method, but it doesn't seem to be documented and I don't know if we need a particular version of wagtail before it's available.

File: cdhweb/pages/migration_utils.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        klass._meta.app_label,
+        klass.__name__,
+        page.url_path
+    )
+
+    return page
+
+# adapted from wagtail's `save_revision`:
+# https://github.com/wagtail/wagtail/blob/stable/2.3.x/wagtail/core/models.py#L631
+def create_revision(apps, page, user=None, created_at=datetime.now(), **kwargs):
```
Comment (rlskoeser) [inline cdhweb/pages/migration_utils.py:71]: Forgot to ask — what do we need revisions for?  Are they required for content to display?

File: cdhweb/pages/migrations/0002_homepage.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    site.root_page = home
+    site.root_page_id = home.id
+    site.save()
+
+    # delete the default welcome page
+    welcome = Page.objects.get(title='Welcome to your new Wagtail site!')
+    welcome.delete()
+
+
+def revert_homepage(apps, schema_editor):
```
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply cdhweb/pages/migrations/0002_homepage.py:44]: Revert logic has to be provided for the `TestMigration` class to work, since it has to know how to rollback the migration it applies in the test. This is also why `streamfield_to_html` exists

File: cdhweb/pages/models.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+from wagtail.images.edit_handlers import ImageChooserPanel
+from wagtail.search import index
+
+from cdhweb.blog.models import BlogPost
+from cdhweb.events.models import Event
+from cdhweb.projects.models import Project
+
+
+#: commonly allowed tags for RichTextBlocks
+RICH_TEXT_TAGS = ['h3', 'h4', 'bold', 'italic', 'link', 'ol', 'ul', 'blockquote']
```
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply cdhweb/pages/models.py:23]: > We probably need more than this for cdhweb, though

Can you expand on this?

File: cdhweb/pages/migrations/0002_homepage.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.27 on 2020-01-08 19:59
+from __future__ import unicode_literals
+
+from django.db import migrations
+from cdhweb.pages.migration_utils import add_child, create_revision
+
+
+def create_homepage(apps, schema_editor):
```
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply cdhweb/pages/migrations/0002_homepage.py:9]: It definitely opens up a lot of possibilities! I think having a dry run mode for the manage command would probably solve most of what the migration tests are trying to determine; that's a great idea.

File: cdhweb/pages/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rlskoeser) [reply cdhweb/pages/models.py:23]: Yeah — I think we might want to do an inventory of the tags that are currently in use and then use then figure out what would be a reasonable set to allow.

Would it make sense for the migration script could have a report mode to summarize what tags are used where? (Maybe some cleanup could be done in advance... 🤔 )

BTW shouldn't this list include `li` ?

File: cdhweb/pages/migration_utils.py (hunk 1 above)
Comment (thatbudakguy) [reply cdhweb/pages/migration_utils.py:14]: It's not available in the context of a migration.

File: cdhweb/pages/migration_utils.py (hunk 3 above)
Comment (thatbudakguy) [reply cdhweb/pages/migration_utils.py:71]: revisions associate text content with the page model; it's a one -> many. pages themselves have no content stored in the database and you need at least one revision in order for a page to have content.

**Comment by thatbudakguy** [comment]
Closing, since we can always refer to the conversation on this PR later. I think our eventual approach will not be along the lines of this PR.
